### DISCUSSION TITLE
Help with multiple image uploads, please ?

### DISCUSSION THREAD

**Discussion by agloolik** [react-page/react-page#1181] [Q&A]
Hello team,
On my customer's site the image plugin loads images to cloudinary. My customer would like to be able to load multiple images in one pass, by selecting them in the initial browsing window. then the images would be uploaded to cloudinary account and finally the links added in the page in the same selected order. Any ideas ? Better than doing it myself, could the react-page team send me a quote for this custom work? By the way the site was made with mui 5, so due to react-page working only in v4, it has both versions 4 and 5 running together,
Keep up the good work !
Thanks and regards.

**Comment by agloolik** [comment]
So, I tried to implement this. I can upload images to cloudinary via a custom plugin, but I guess the onChange event, I can't figure out how I should use it to update the renderer here's my code: ( I apoligize for this ugly coding style, right now I'm just trying to get this to work :)
The Plugin code:

```tsx
import { CellPlugin } from "@react-page/editor";
import ImageUploadWidget from "../components/admin/ImageUploadWidget";
import Image from "next/image";
import { lazyLoad } from "@react-page/editor";

// use a type here, not an interface
type CloudinaryImage = {
  asset_id: string;
  height: number;
  width: number;
  secure_url: string;
};
const iconStyle: React.CSSProperties = {
  width: "100%",
  height: "auto",
  padding: "0",
  color: "#aaa",
  textAlign: "center",
  minWidth: 64,
  minHeight: 64,
  maxHeight: 256,
};

// type Data = { images: CloudinaryImage[] };
const ImageIcon = lazyLoad(() => import("@mui/icons-material/Landscape"));

const cloudinaryUploads: CellPlugin<{ images: CloudinaryImage[] }> = {
  Renderer: ({ data }) => (
    <>
      {data.images ? (
        data.images.map((image) => (
          <Image
            className={"image"}
            key={image.asset_id}
            src={image.secure_url}
            height={image.height}
            width={image.width}
            layout="responsive"
          />
        ))
      ) : (
        <div>
          <div className="react-page-plugins-content-image-placeholder">
            <ImageIcon style={iconStyle} />
          </div>
        </div>
      )}
    </>
  ),
  id: "tasCloudinaryUpload",
  version: 1,
  title: "Cloudinary upload",
  description: "Upload multiple images at once to cloudinary",
  controls: {
    type: "custom",
    Component: (props) => <ImageUploadWidget {...props} />,
  },
};
export default cloudinaryUploads;

The custom component aka controls: {type: "custom" ...}


import { useEffect, useState } from "react";
import { generateSignature } from "../../lib/generateSignature";
import Button from "@mui/material/Button";
import { cloudinary } from "../../config";

const UploadWidget = (props) => {
  const { CLOUDINARY_CLOUD_NAME, CLOUDINARY_KEY, CLOUDINARY_FOLDER } =
    cloudinary;
  const [images, setImages] = useState([]);
  const addImage = (image) => {
    setImages((prevImages) => [...prevImages, image]);
  };

  console.log({ props: props });
  useEffect(() => {
    console.log({ images: images });
    props.onChange(images);
  }, [images]);

  async function handleWidgetClick() {
    const widget = window.cloudinary.createUploadWidget(
      {
        cloudName: CLOUDINARY_CLOUD_NAME,
        uploadSignature: generateSignature,
        apiKey: CLOUDINARY_KEY,
        folder: CLOUDINARY_FOLDER,
        resourceType: "image",
      },
      (error, result) => {
        if (!error && result && result.event === "success") {
          console.log("Done! Here is the image info: ", result.info);
          addImage(result.info);
        } else if (error) {
          console.log(error);
        }
      }
    );
    console.log(props);
    widget.open();
  }

  return <Button onClick={handleWidgetClick}>Upload image</Button>;
};

export default UploadWidget;
```

**Reply by macrozone** [reply]
looks conceptually right to me, alltough i don't know about the cloudinary widget.

There are surely some flaws with it, but does it work already? What is the actual problem?

**Reply by macrozone** [reply]
some things:

- i recommend to use typescript and annotate your props of the "UploadWidget" component properly.  That would have revealed the problem.
- you use a complete custom control for the whole plugin. Consider just using a custom input field only for the `images` property: https://react-page.github.io/docs/#/custom-cell-plugins?id=custom-field-components-example
- no matter if you use a full custom controls or a custom field, `value` and `onChange` should be "symetric": if `value` has type `T`, then `onChange` should pass a new value of type `T`

e.g. if you use a custom component for controls like in your example, that `T` is the shape of the data of your cell plugin. in your case its

```
{
  images: CloudinaryImage[]
}
```

so your `onChange` has to pass the full object: `onChange({ images: newimages } )`

If you would use a custom field however, your input component only deals with the `images` prop, so it should pass `onChange(newImages)`


- in both cases you also have to respect the incoming `values` prop. Custom fields and controls should be "controlled" (and therefore (mostly) stateless). So you should get rid of `  const [images, setImages] = useState([]);`

**Comment by agloolik** [comment]
Any help would be very appreciated, oh and congrats on v5 !

**Comment by macrozone** [comment]
(please use backticks to format code)

**Comment by agloolik** [comment]
Hi, it works in that it shows the button for the cloudinary widget in the controls area. the widget works and upload images to the cloud.
Then the state gets updated in the custon component "UploadWidget" but the images never appear in the renderer.
So they never appear on the page.
Thanks !

**Comment by agloolik** [comment]
@macrozone  : Thank you SO MUCH !!!
I am progressively moving to typescript, the learning curve is a bit steep for me though...
